src/NFT/AgentCommunication.sol

```diff
@@ -4,19 +4,70 @@ pragma solidity ^0.8.22;
 import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
 import "./DoubleEndedStructQueue.sol";
 
-contract AgentCommunication is Ownable {
-    address payable public treasury;
-    uint256 public pctToTreasuryInBasisPoints; //70% becomes 7000
+interface IAgentRegistry {
+    error AgentNotRegistered();
+
+    event AgentRegistered(address indexed agent);
+    event AgentDeregistered(address indexed agent);
+
+    function registerAsAgent() external;
+    function deregisterAsAgent() external;
+    function isRegisteredAgent(address agent) external view returns (bool);
+    function getAllRegisteredAgents() external view returns (address[] memory);
+}
+
+contract AgentRegistry is IAgentRegistry, Ownable {
+    mapping(address => bool) public registeredAgents;
+    address[] private registeredAgentsList;
```

Comment:
Consider using EnumerableMap (https://docs.openzeppelin.com/contracts/5.x/api/utils#EnumerableMap) instead of mapping + list, more efficient.